fix(ngcc): recognize re-exports with imported TS helpers in CommonJS and UMD

gkalpak opened the pull request:
~~_Builds on top of #34512 (so the first 7 commits are from there)._~~
_Rebased on master now that #34512 has been merged._

##
Previously, the `CommonJsReflectionHost` and `UmdReflectionHost` would only recognize re-exports of the form `__export(...)`. This is what re-exports look like, when the TypeScript helpers are emitted inline (i.e. when compiling with the default [TypeScript compiler options][1] that include `noEmitHelpers: false` and `importHelpers: false`).

However, when compiling with `importHelpers: true` and [tslib][2] (which is the recommended way for optimized bundles), the re-exports will look like: `tslib_1.__exportStar(..., exports)`
These types of re-exports were previously not recognized by the CommonJS/UMD `ReflectionHost`s and thus ignored.

This commit fixes this by ensuring both re-export formats are recognized.

[1]: https://www.typescriptlang.org/docs/handbook/compiler-options.html
[2]: https://www.npmjs.com/package/tslib

gkalpak committed:
fix(ngcc): recognize re-exports with imported TS helpers in CommonJS and UMD

Previously, the `CommonJsReflectionHost` and `UmdReflectionHost` would
only recognize re-exports of the form `__export(...)`. This is what
re-exports look like, when the TypeScript helpers are emitted inline
(i.e. when compiling with the default [TypeScript compiler options][1]
that include `noEmitHelpers: false` and `importHelpers: false`).

However, when compiling with `importHelpers: true` and [tslib][2] (which
is the recommended way for optimized bundles), the re-exports will look
like: `tslib_1.__exportStar(..., exports)`
These types of re-exports were previously not recognized by the
CommonJS/UMD `ReflectionHost`s and thus ignored.

This commit fixes this by ensuring both re-export formats are
recognized.

[1]: https://www.typescriptlang.org/docs/handbook/compiler-options.html
[2]: https://www.npmjs.com/package/tslib

mary-poppins commented:
You can preview d66da74 at https://pr34527-d66da74.ngbuilds.io/.

gkalpak committed:
fixup! fix(ngcc): recognize re-exports with imported TS helpers in CommonJS and UMD

petebacondarwin commented on packages/compiler-cli/ngcc/test/host/commonjs_host_spec.ts:
```diff
@@ -554,13 +555,22 @@ exports.xtra2 = xtra2;
         {
           name: _('/wildcard_reexports.js'),
```
Perhaps rename this file to be `wildcard_reexports_inline_helpers.js`?

gkalpak replied:
:+1: I'll go with `*_emitted_helpers` (see comment below).

petebacondarwin commented on packages/compiler-cli/ngcc/test/host/commonjs_host_spec.ts:
```diff
@@ -1763,7 +1773,7 @@ exports.ExternalModule = ExternalModule;
         });
 
-        it('should handle wildcard re-exports of other modules', () => {
+        it('should handle wildcard re-exports of other modules (with emitted helpers)', () => {
```
NIT: emitted -> inline

(Unless I am wrong about the convention for naming these.)

gkalpak replied:
Both `emitted` and `inline` coule be used (the helpers are emitted inline by TS). I'll keep `emitted` to avoid confusion with `inline exports` (which is something different).

petebacondarwin reviewed (approved):
LGTM (for last commit) - couple of minor nits.

gkalpak commented:
Thx for the review, @petebacondarwin! Rebased on master and addressed feedback.
Marking this as blocked on #34512 (to ensure they are merged in order).

mary-poppins commented:
You can preview 9fbc8ce at https://pr34527-9fbc8ce.ngbuilds.io/.

gkalpak commented:
Rebased on master.

mary-poppins commented:
You can preview dda29a4 at https://pr34527-dda29a4.ngbuilds.io/.

petebacondarwin commented:
CI?

mary-poppins commented:
You can preview 841ecf9 at https://pr34527-841ecf9.ngbuilds.io/.

gkalpak commented:
#34512 has been merged. Rebased on master and marked for merging.

mary-poppins commented:
You can preview 13d3dab at https://pr34527-13d3dab.ngbuilds.io/.
